statsmodels/stats/multitest.py

```diff
@@ -444,3 +444,237 @@ def fdrcorrection_twostage(pvals, alpha=0.05, method='bky', iter=False,
+        The estimated mean of the empirical null distribution
+    scale0 : float
+        The estimated standard deviation of the empirical null distribution
+    prob0 : float
+        The estimated proportion of true null hypotheses
+    null_density : function (from reals to positive reals)
+        The fitted empirical null density function
+
+    References
+    ----------
+    B Efron (2008).  Microarrays, Empirical Bayes, and the Two-Groups
+    Model.  Statistical Science 23:1, 1-22.
+
+    Notes
+    -----
+    See also:
+
+    http://nipy.org/nipy/labs/enn.html#nipy.algorithms.statistics.empirical_pvalue.NormalEmpiricalNull.fdr
+    """
+
+    def __init__(self, zscores, null_lb=-1, null_ub=1, estimate_mean=True,
+                 estimate_scale=True, estimate_prob=False):
+
+        # Extract the null z-scores
+        ii = np.flatnonzero((zscores >= null_lb) & (zscores <= null_ub))
+        if len(ii) == 0:
+            raise RunTimeError("No Z-scores fall between null_lb and null_ub")
+        zscores0 = zscores[ii]
+
+        # Number of Z-scores, and null Z-scores
+        n_zs, n_zs0 = len(zscores), len(zscores0)
+
+        # Unpack and transform the parameters to the natural scale, hold
+        # parameters fixed as specified.
+        def xform(params):
+
+            mean0, logscale0, logitp0 = tuple(params)
+            scale0 = np.exp(logscale0)
+            prob0 = 1 / (1 + np.exp(-logitp0))
+
+            mean0 = mean0 if estimate_mean else 0.
+            scale0 = scale0 if estimate_scale else 1.
+            prob0 = prob0 if estimate_prob else 1.
+
+            return mean0, scale0, prob0
+
+
+        from scipy.stats.distributions import norm
+
+
+        def fun(params):
+            """
+            Negative log-likelihood of z-scores.
+
+            The function has three arguments, packed into a vector:
+
+            mean : location parameter
+            logscale : log of the scale parameter
+            logitprop : logit of the proportion of true nulls
+
+            The implementation follows section 4 from Efron 2008.
+            """
+
+            d0, s0, p0 = xform(params)
+
+            # Mass within the central region
+            central_mass = (norm.cdf((null_ub - d0) / s0) -
+                            norm.cdf((null_lb - d0) / s0))
+
+            # Probability that a Z-score is null and is in the central region
+            cp = p0 * central_mass
+
+            # Binomial term
+            rval = n_zs0 * np.log(cp) + (n_zs - n_zs0) * np.log(1 - cp)
+
+            # Truncated Gaussian term for null Z-scores
+            zv = (zscores0 - d0) / s0
+            rval += np.sum(-zv**2 / 2) - n_zs0 * np.log(s0)
+            rval -= n_zs0 * np.log(central_mass)
+
+            return -rval
+
+
+        # Estimate the parameters
+        from scipy.optimize import minimize
+        # starting values are mean = 0, scale = 1, p0 ~ 1
+        mz = minimize(fun, np.r_[0., 0, 3], method="Nelder-Mead")
```

Comment:
the minimization uses full params even if some are fixed. It might not be a problem in this simple optimization problem, but should have a unit test (if mean, variance or both are fixed).